model-providers/ollama/runtime/src/main/java/io/quarkiverse/langchain4j/ollama/OllamaDefaultToolsHandler.java

```diff
@@ -0,0 +1,132 @@
+
+            When the user provides sufficient information , answer with the __conversational_response tool.
+            """);
+
+    static final ToolSpecification DEFAULT_RESPONSE_TOOL = ToolSpecification.builder()
+            .name("__conversational_response")
+            .description("Respond conversationally if no other tools should be called for a given query and history.")
+            .parameters(ToolParameters.builder()
+                    .type("object")
+                    .properties(
+                            Map.of("reponse",
+                                    Map.of("type", "string",
+                                            "description", "Conversational response to the user.")))
+                    .required(Collections.singletonList("response"))
+                    .build())
+            .build();
+
+    @Override
+    public Builder enhanceWithTools(Builder builder, List<Message> messages, List<ToolSpecification> toolSpecifications,
+            ToolSpecification toolThatMustBeExecuted) {
+        if (toolSpecifications == null || toolSpecifications.isEmpty()) {
+            return builder;
+        }
+        // Set Json Format
+        builder.format("json");
+
+        // Construct prompt with tools toolSpecifications and DEFAULT_RESPONSE_TOOL
+        List<ToolSpecification> extendedList = new ArrayList<>(toolSpecifications.size() + 1);
+        extendedList.addAll(toolSpecifications);
+        extendedList.add(DEFAULT_RESPONSE_TOOL);
+        Prompt prompt = DEFAULT_SYSTEM_TEMPLATE.apply(
+                Map.of("tools", Json.toJson(extendedList)));
+
+        // TODO handle -> toolThatMustBeExecuted skipped for the moment
+        String initialSystemMessages = messages.stream().filter(cm -> cm.role() == Role.SYSTEM)
+                .map(Message::content)
+                .collect(Collectors.joining("\n"));
+
+        Message groupedSystemMessage = Message.builder()
+                .role(Role.SYSTEM)
+                .content(initialSystemMessages + "\n" + prompt.text())
+                .build();
+
+        List<Message> otherMessages = messages.stream().filter(cm -> cm.role() != Role.SYSTEM).toList();
+
+        // Add specific tools message
+        List<Message> messagesWithTools = new ArrayList<>(messages.size() + 1);
+        messagesWithTools.add(groupedSystemMessage);
+        messagesWithTools.addAll(otherMessages);
+
+        builder.messages(messagesWithTools);
+
+        return builder;
+    }
+
+    @Override
+    public AiMessage getAiMessageFromResponse(ChatResponse response, List<ToolSpecification> toolSpecifications) {
+        ToolResponse toolResponse;
+        try {
+            // Extract tools
+            toolResponse = Json.fromJson(response.message().content(), ToolResponse.class);
+        } catch (Exception e) {
+            // No tools found
+            return AiMessage.from(response.message().content());
+        }
```

Comment:
I'm not a fan using for exceptions for control flow. Can we find another way of figuring out whether or not tools are needed?

Reply:
It's really an exception, Ollama server should always answer with a json doc corresponding to ToolResponse.
I've change the behavior to throw an exception.